src/modalities/models/lora/utils.py

```diff
@@ -0,0 +1,112 @@
+import copy
+
+from torch import nn
+
+from modalities.models.lora.lora_layers import LoRAEmbedding, LoRALinear, LoRALayer
+
+
+def mark_only_lora_as_trainable(model: nn.Module, bias: str = "none") -> None:
+    for n, p in model.named_parameters():
+        if "lora_" not in n:
+            p.requires_grad = False
+    if bias == "none":
+        return
+    elif bias == "all":
+        for n, p in model.named_parameters():
+            if "bias" in n:
+                p.requires_grad = True
+    elif bias == "lora_only":
+        for m in model.modules():
+            if isinstance(m, LoRALayer) and hasattr(m, "bias") and m.bias is not None:
+                m.bias.requires_grad = True
+    else:
+        raise NotImplementedError
+
+
+def conversion_lora(
+    model: nn.Module,
+    r: int,
+    alpha: int,
+):
+    replace_modules_in_attention(model=model, r=r, alpha=alpha)
+    mark_only_lora_as_trainable(model=model)
+    return model
+
+def replace_modules_in_attention(
+    model: nn.Module,
+    r: int,
+    alpha: int,
+):
+    # todo implement with different layer types
+    # "attn", "linear", "embedding", "conv1d", "conv2d", "conv3d"
+    # also implement with different model key name 'attn'
+    for name, module in model.named_children():
+        if "attention" in type(module).__name__.lower() and isinstance(module, nn.Module):
+            for sub_name, sub_module in module.named_children():
+                if isinstance(sub_module, nn.Linear) or isinstance(sub_module, nn.Embedding):
+                    new_linear = convert_layer(sub_module, r, alpha)
+                    setattr(module, sub_name, new_linear)
+        else:
+            replace_modules_in_attention(module, r, alpha)
+
+def convert_layer(layer: nn.Module, r: int, alpha: int) -> nn.Module:
+    if isinstance(layer, nn.Embedding):
+        result = convert_embedding(layer, r, alpha)
+    elif isinstance(layer, nn.Linear):
+        result = convert_linear(layer, r, alpha)
+    else:
+        # todo handle this
```

Comment:
This we still need to think about: 
Suggestion: we log that we wanted to convert the layer but didn't and do not raise an exception. 
Also, we need to add the conv1d / 2d / 3d layers here.